File: rfhe_framewk/src/baseConv.cpp
// baseConv.cpp
#include <iostream>
#include <vector>
#include <random>
#include <algorithm>
#include <numeric>
#include <tuple>
#include <set>
#include <cstdint>
#include <climits>

// 128-bit 整数别名
using uint128 = unsigned __int128;
using uint64 = uint64_t;

// --- Miller-Rabin 素性测试（确定性，适用于 64 位） ---
uint64 mod_pow(uint64 a, uint64 d, uint64 m) {
    uint128 res = 1, base = a % m;
    while (d) {
        if (d & 1) res = (res * base) % m;
        base = (uint128)base * base % m;
        d >>= 1;
    }
    return (uint64)res;
}

bool is_prime(uint64 n) {
    if (n < 2) return false;
    for (uint64 p : {2ULL, 3ULL, 5ULL, 7ULL, 11ULL, 13ULL, 17ULL, 19ULL, 23ULL}) {
        if (n % p == 0) return n == p;
    }
    uint64 d = n - 1, s = 0;
    while ((d & 1) == 0) { d >>= 1; ++s; }
    for (uint64 a : {2ULL, 325ULL, 9375ULL, 28178ULL, 450775ULL, 9780504ULL, 1795265022ULL}) {
        if (a % n == 0) continue;
        uint64 x = mod_pow(a, d, n);
        if (x == 1 || x == n - 1) continue;
        bool comp = true;
        for (uint64 r = 1; r < s; ++r) {
            x = (uint128)x * x % n;
            if (x == n - 1) { comp = false; break; }
        }
        if (comp) return false;
    }
    return true;
}

uint64 next_prime(uint64 cand) {
    if ((cand & 1) == 0) ++cand;
    while (!is_prime(cand)) cand += 2;
    return cand;
}

// 扩展欧几里得，计算 a * x + m * y = gcd(a, m)
int64_t egcd(int64_t a, int64_t b, int64_t &x, int64_t &y) {
    if (b == 0) { x = 1; y = 0; return a; }
    int64_t x1, y1;
    int64_t g = egcd(b, a % b, x1, y1);
    x = y1;
    y = x1 - (a / b) * y1;
    return g;
}

// 计算 a 在模 m 下的乘法逆元
uint64 mod_inv(uint64 a, uint64 m) {
    int64_t x, y;
    int64_t g = egcd(a, m, x, y);
    if (g != 1) throw std::runtime_error("No inverse");
    x %= (int64_t)m;
    if (x < 0) x += m;
    return (uint64)x;
}

// 生成 limbs 个 bitwidth 位素数
std::vector<uint64> generate_crt_primes(int limbs, int bitwidth) {
    if (bitwidth < 10) throw std::invalid_argument("bitwidth must be at least 10");
    std::mt19937_64 gen(std::random_device{}());
    std::set<uint64> seen;
    std::vector<uint64> primes;
    uint64 mask = (bitwidth == 64 ? ~0ULL : ((1ULL << bitwidth) - 1));
    while ((int)primes.size() < limbs) {
        uint64 cand = (gen() & mask) | (1ULL << (bitwidth - 1)) | 1ULL;
        uint64 p = next_prime(cand);
        int bl = p ? 64 - __builtin_clzll(p) : 1;
        if (bl == bitwidth && seen.insert(p).second) {
            primes.push_back(p);
        }
    }
    return primes;
}

// 生成 poly_dim + 1 长度的残余向量（最后一位为 ECC 校验和）
std::vector<std::vector<uint64>> generate_residues(
    const std::vector<uint64> &moduli, int poly_dim)
{
    std::mt19937_64 gen(std::random_device{}());
    std::vector<std::vector<uint64>> residues;
    for (auto p : moduli) {
        std::uniform_int_distribution<uint64> dist(0, p - 1);
        std::vector<uint64> arr(poly_dim + 1);
        uint64 sum = 0;
        for (int i = 0; i < poly_dim; ++i) {
            arr[i] = dist(gen);
            sum = (sum + arr[i]) % p;
        }
        arr[poly_dim] = sum;
        residues.push_back(arr);
    }
    return residues;
}

// 随机在矩阵中选 m 个元素，然后总共翻转 n 位
void flip_n_bits_across_m_elements_2d_inplace(
    std::vector<std::vector<uint64>> &matrix,
    int m, int n)
{
    int rows = matrix.size();
    int cols = matrix[0].size();
    int total = rows * cols;
    if (m < 1 || m > total) throw std::invalid_argument("Invalid m");
    std::vector<std::pair<int,int>> pos;
    pos.reserve(total);
    for (int i = 0; i < rows; ++i)
        for (int j = 0; j < cols; ++j)
            pos.emplace_back(i, j);
    std::mt19937_64 gen(std::random_device{}());
    std::shuffle(pos.begin(), pos.end(), gen);
    pos.resize(m);

    std::vector<std::tuple<int,int,int>> bits;
    for (auto [r,c] : pos) {
        uint64 v = matrix[r][c];
        int w = v ? 64 - __builtin_clzll(v) : 1;
        for (int b = 0; b < w; ++b)
            bits.emplace_back(r, c, b);
    }
    if (n < 0 || n > (int)bits.size()) throw std::invalid_argument("Invalid n");
    std::shuffle(bits.begin(), bits.end(), gen);
    for (int i = 0; i < n; ++i) {
        auto [r,c,b] = bits[i];
        matrix[r][c] ^= (1ULL << b);
    }
}

// CRT 重建（返回长度为 N 的向量，使用 128 位运算避免溢出）
std::vector<uint128> crt_reconstruct(
    const std::vector<std::vector<uint64>> &residues,
    const std::vector<uint64> &moduli)
{
    int m = moduli.size();
    int N = residues[0].size();
    uint128 P = 1;
    for (auto p : moduli) P *= p;

    std::vector<uint128> hat_p(m);
    std::vector<uint64> inv_hat_p(m);
    for (int j = 0; j < m; ++j) {
        hat_p[j] = P / moduli[j];
        inv_hat_p[j] = mod_inv((uint64)(hat_p[j] % moduli[j]), moduli[j]);
    }

    std::vector<uint128> result(N);
    for (int i = 0; i < N; ++i) {
        uint128 x = 0;
        for (int j = 0; j < m; ++j) {
            uint128 term = (uint128)residues[j][i] * hat_p[j] % P;
            term = term * inv_hat_p[j] % P;
            x = (x + term) % P;
        }
        result[i] = x;
    }
    return result;
}

int main() {
    const int limbs         = 4;
    const int crt_bitwidth  = 20;
    const int poly_dim      = 1024;
    const int flip_bits     = 2;
    const int flip_elements = 2;
    const int epochs        = 100000;
    const int bar_width     = 50;

    uint64 ecc_pass = 0, ecc_fail = 0, crt_fail = 0;

    for (int e = 0; e < epochs; ++e) {
        // 进度条
        if (e % (epochs / bar_width == 0 ? 1 : epochs / bar_width) == 0) {
            int filled = e * bar_width / epochs;
            std::cout << "\r[";
            for (int i = 0; i < bar_width; ++i) std::cout << (i < filled ? '=' : ' ');
            std::cout << "] " << (e * 100 / epochs) << "%" << std::flush;
        }

        auto moduli   = generate_crt_primes(limbs, crt_bitwidth);
        auto residues = generate_residues(moduli, poly_dim);
        auto original = residues;
        flip_n_bits_across_m_elements_2d_inplace(residues, flip_elements, flip_bits);

        auto recon = crt_reconstruct(residues, moduli);

        bool crt_err = false;
        for (int i = 0; i < (int)recon.size(); ++i) {
            for (int j = 0; j < limbs; ++j) {
                if ((recon[i] % moduli[j]) != original[j][i]) {
                    crt_err = true; break;
                }
            }
            if (crt_err) break;
        }
        if (crt_err) ++crt_fail;

        uint128 P = 1; for (auto p : moduli) P *= p;
        uint128 sum = 0; for (int i = 0; i < poly_dim; ++i) sum = (sum + recon[i]) % P;
        if (sum == recon[poly_dim]) ++ecc_pass; else ++ecc_fail;
    }
    std::cout << "\r[" << std::string(bar_width, '=') << "] 100%\n";

    std::cout << "EPOCHS: "        << epochs    << "\n"
              << "ECC FAIL TIME: " << ecc_fail  << "\n"
              << "CRT FAIL TIME: " << crt_fail  << "\n";
    return 0;
}

File: rfhe_framewk/src/baseConv.py
from typing import List
import random
import math
from sympy import nextprime
import copy
from tqdm import tqdm
import numpy as np


def bConv(
    residue_arrays: List[List[int]], moduli: List[int], moduli_out: List[int]
) -> List[int]:
    m = len(moduli)
    N = len(residue_arrays[0])
    P = math.prod(moduli)

    hat_p = [P // p for p in moduli]
    inv_hat_p = [pow(hat_p[j], -1, moduli[j]) for j in range(m)]

    result = []
    for i in range(N):
        x = 0
        main_limb = []
        for j in range(m):
            sub_limb = []
            for q in moduli_out:
                r = int(residue_arrays[j][i])
                term = (r * hat_p[j] * inv_hat_p[j])
                sub_limb.append(term % q)
            main_limb.append(sub_limb)

        summed = []
        for k in range(len(moduli_out)):
            total = 0
            for j in range(m):
                total += main_limb[j][k]
            summed.append(total)
        result.append(summed)
    # print(np.array(result))
    return result

def bConv_ecc(
    residue_arrays: List[List[int]], moduli: List[int], moduli_out: List[int]
) -> List[int]:
    m = len(moduli)
    P = math.prod(moduli)

    for i in range(m):
        residue_arrays[i].append(sum(residue_arrays[i]) % moduli[i])

    N = len(residue_arrays[0])

    hat_p = [P // p for p in moduli]
    inv_hat_p = [pow(hat_p[j], -1, moduli[j]) for j in range(m)]

    result = []
    full_data = []
    for i in range(N):
        x = 0
        main_limb = []
        for j in range(m):
            sub_limb = []
            for q in moduli_out:
                r = int(residue_arrays[j][i])
                term = (r * hat_p[j] * inv_hat_p[j])
                sub_limb.append(term % q)
            # print(sum(sub_limb[:-1]), sub_limb[-1])
            main_limb.append(sub_limb)
        full_data.append(main_limb)

        summed = []
        for k in range(len(moduli_out)):
            total = 0
            for j in range(m):
                total += main_limb[j][k]
            summed.append(total)
        result.append(summed)
    # —— 在 bConv_ecc 中，累加前 N‑1 项时做模 q ——  
    sum_data = [[0]*len(moduli_out) for _ in range(m)]
    for item in full_data[:-1]:
        for j in range(m):
            for k, q in enumerate(moduli_out):
                # 每步都 mod q，保持在 [0..q-1]
                sum_data[j][k] = (sum_data[j][k] + item[j][k]) % q

    # ECC 项本身就是 mod q 后的
    ecc_item = full_data[-1]

    # 再比较
    for j in range(m):
        for k, q in enumerate(moduli_out):
            if sum_data[j][k] != ecc_item[j][k]:
                raise ValueError(
                    f"ECC mismatch at j={j}, k={k}: "
                    f"sum_mod={sum_data[j][k]}, ecc={ecc_item[j][k]}"
                )

    return result

def generate_crt_primes(limbs: int, bitwidth: int) -> List[int]:
    if bitwidth < 10:
        raise ValueError("bitwidth must be at least 10")
    primes = []
    seen = set()
    while len(primes) < limbs:
        cand = random.getrandbits(bitwidth) | (1 << (bitwidth - 1)) | 1
        p = int(nextprime(cand))
        if p.bit_length() == bitwidth and p not in seen:
            primes.append(p)
            seen.add(p)
    return primes


def generate_residues(moduli: List[int], poly_dim: int) -> List[List[int]]:
    residues = []
    for p in moduli:
        arr = [random.randrange(p) for _ in range(poly_dim)]
        residues.append(arr)
    return residues


import random
from typing import List, Optional, Tuple


def flip_n_bits_across_m_elements_2d_inplace(
    matrix: List[List[int]],
    m: int,
    n: int,
    bit_width: Optional[int] = None,
    max_bit: Optional[int] = None,
) -> None:
    """
    在二维整数矩阵中，随机选 m 个不同元素，然后在这 m 个元素的所有位位置里
    随机翻转总共 n 个位（直接在原矩阵上修改，无返回值）。

    Args:
        matrix:   二维列表，matrix[row][col] 是一个整数。
        m:        要选取的元素个数，必须 1 <= m <= 总元素数。
        n:        要翻转的总位数，必须 0 <= n <= m * 最大位宽。
        bit_width:
                  如果指定，则所有元素都按这个固定的位宽来考虑位位置 [0..bit_width-1]；
                  否则按元素自身的 bit_length()（至少 1）。
        max_bit:
                  如果指定，则只允许翻转位索引 <= max_bit 的位，
                  也就是最终可选 bitpos 范围会被截断到 [0..max_bit]。

    Raises:
        ValueError: 当 m 或 n 超出合理范围时。
    """
    # 扁平化所有元素的位置
    all_pos = [(r, c) for r in range(len(matrix)) for c in range(len(matrix[r]))]
    total = len(all_pos)
    if not (1 <= m <= total):
        raise ValueError(f"m must be between 1 and {total}, got {m}")

    # 随机选 m 个不同元素
    chosen = random.sample(all_pos, m)

    # 构造所有 (r, c, bitpos) 候选
    cand_bits: List[Tuple[int, int, int]] = []
    for r, c in chosen:
        val = matrix[r][c]
        # 确定位宽
        if bit_width is not None:
            w = bit_width
        else:
            w = val.bit_length() or 1
        # 如果 max_bit 限制了最高可翻转位，则截断
        if max_bit is not None:
            w = min(w, max_bit + 1)
        for b in range(w):
            cand_bits.append((r, c, b))

    if not (0 <= n <= len(cand_bits)):
        raise ValueError(f"n must be between 0 and {len(cand_bits)}, got {n}")

    # 随机选 n 个 (r, c, b) 并翻转
    for r, c, b in random.sample(cand_bits, n):
        matrix[r][c] ^= 1 << b


def group_multiply(lst, n):
    result = []
    for i in range(0, len(lst), n):
        group = lst[i : i + n]
        if len(group) < n:
            break
        product = 1
        for x in group:
            product *= x
        result.append(product)
    return result


if __name__ == "__main__":
    limbs, limbs_out, crt_bitwidth, poly_dim = 4, 2, 10, 1024
    ecc_pass = 0
    ecc_fail = 0
    crt_fail = 0
    flip_bits = 2
    flip_elements = 2
    epoches = 1
    for _ in tqdm(range(epoches), desc="Epochs", unit="epoch"):
        moduli = generate_crt_primes(limbs, crt_bitwidth)
        residues = generate_residues(moduli, poly_dim)
        # print(residues)
        # print(moduli)
        residues_original = copy.deepcopy(residues)
        moduli_out = group_multiply(moduli, limbs // limbs_out)
        # flip_n_bits_across_m_elements_2d_inplace(residues, flip_elements, flip_bits,
        #                                           # max_bit=crt_bitwidth
        #                                       )
        recon = bConv_ecc(residues, moduli, moduli_out)

